### DISCUSSION TITLE
How to implement Callbacks in Fusion language?

### DISCUSSION THREAD

**Discussion by ghost** [pfusik/cito#127] [Q&A] [answered]
I have read the reference but I can't find anything about it. Please help. Thank you.

**Comment by pfusik** [comment]
```csharp
public abstract class MyClass
{
    protected abstract void MyCallback();
}
```

or:

```csharp
public class MyClass
{
    protected virtual void MyCallback()
    {
    }
}
```

and subclass in your target language.

**Reply by ghost** [reply]
Are we talking about the same `Callbacks`? I'm asking about something similar to delegates in C#, function pointers in C and `std::function` in C++.

**Reply by pfusik** [reply]
Yes. This is designed Java-style through subclassing. Are you familiar with Java?

C/C++ function pointers are less powerful, because they contain no object reference.

**Reply by ghost** [reply]
I'm not familiar to Java style Callbacks.

**Reply by pfusik** [reply]
Basically, it's just virtual methods as I wrote above. Do you understand virtual methods?

**Reply by pfusik** [reply]
We might have some misunderstanding because I don't know your exact use case. The callback can be defined in your target language, but also in Fusion.

**Reply by BenBE** [reply]
> C/C++ function pointers are less powerful, because they contain no object reference.

I need to object:

- For C the common idiom is providing a `void* ptr` for user data, which is in turn used to reference some "object" …

- For C++: Use `std::function` and use lambda functions with context binding:

```cpp
std::function<void()> cb_function = [this]() {
    this->matter();
};
```

Or use the variant from C. Might be surprising, but that also works on C++ …

The Java-style callbacks are just plain BS, because they force arguments when none would be needed otherwise.

**Reply by pfusik** [reply]
I was referring to function pointers themselves. What you describe is a pair of a function pointer and a data pointer. That is what initial versions of Fusion used.

> they force arguments when none would be needed otherwise

I don't understand what are you referring to here?

**Reply by BenBE** [reply]
> > they force arguments when none would be needed otherwise
> 
> I don't understand what are you referring to here?

The Java-style callbacks always carry an implicit reference to the object they are created on. Thus you always have to provide for this extra, implicit argument in the underlying target language even in cases where you would not normally need it (e.g. I just need some function called once an event takes place) or where explicit placement needs to be ensured (e.g. most `Enum*` APIs on Windows).

And while you can't pass `std::function` to those APIs directly either, there's normally a few lines of code you can auto-generate to create a matching wrapper (which wouldn't even hurt in the case of fut as you auto-generate code anyway). That aside, IIRC there's even a macro/API available on Windows, which creates this wrapper at runtime …

**Reply by ghost** [reply]
I don't know that much in the detail. But I have the same opinion. Java-style callbacks are BS. Poor Java doesn't support function pointers so it's impossible to pass a method as parameter to another method. Java developers, which so keen on design patterns, created this own design pattern to overcome Java's drawback. It's an incomplete patchwork. I don't know why you praised it.

C# nowadays is a bloated unfathomable mess. But please recall that it used to be a more reasonable version of Java. It supports delegates from the beginning. You decided to follow C# syntax, but adopted Java's no unsigned types and Java's poor workaround for callbacks. I have no idea.

**Reply by pfusik** [reply]
@BenBE Calling a function with an object reference vs no object reference is a difference of one CPU instruction. Unrelevant for performance. Easily optimizable with LTO. What's more, my experience is that most callbacks do need their context data.

@iahung2 I'm not going to join this Java-vs-C# flamewar, especially that you admitted to not know Java.

Both: Could you please show your code snippet so we can discuss if any Fusion syntax improvements would make a considerable difference?

**Comment by pfusik** [comment, marked as answer]
Another example:
```csharp
public abstract class ThreeIntsFull
{
    protected abstract void OnFull!();
}

public class ThreeIntsContainer
{
    int[3] Content;
    int Count = 0;

    public void Append(int value, ThreeIntsFull! fullCallback)
    {
        if (Count >= 3)
            fullCallback.OnFull();
        else
            Content[Count++] = value;
    }
}
```

or:

```csharp
public abstract class ThreeIntsContainer
{
    int[3] Content;
    int Count = 0;
    protected abstract void OnFull!();

    public void Append(int value)
    {
        if (Count >= 3)
            OnFull();
        else
            Content[Count++] = value;
    }
}
```
